Require west 0.6.0

mbolivar opened the pull request:
We're starting to cut west 0.6.0 here: https://github.com/zephyrproject-rtos/west/pull/282

The release needs documentation. There are incompatible API changes from v0.5.x, so I can't do that conveniently while still allowing west 0.5.x in zephyr, since the west docs are currently bundled with the Zephyr docs.

So bump the minimum west version and add documentation for the v0.6.x series.

Not mergeable until west 0.6 is out (and until our CI container is updated to install it), but putting this out there for review.

mbolivar-nordic committed:
doc: add a release notes page for west

This just documents what's already happened.

A subsequent patch will extend it for v0.6.x.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: move runner documentation out of west-apis.rst

Now that the runners package no longer depends on west, we can move
the documentation for it from the orphan west-apis.rst page to the
build-flash-debug.rst page that documents the west extension commands
which use that package.

(This is a more logical place for this information, but it was
previously not possible to put it there since we must be able to build
the documentation on a system without west installed, excepting
content in west-apis.rst. Removing the dependency means the runners
automodule directive can safely appear in build-flash-debug.rst.)

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
west: require v0.6.0 or higher

The main change is the elimination of the bootstrapper, a design flaw
/ misfeature.

Update the documentation to be compatible with the 0.6.x releases as
well. This has to be done atomically, as there were incompatible
changes.  Make use of the versionchanged and versionadded directives
to begin keeping track of how these APIs are evolving.

(Note that west 0.6.0 will remain compatible with the extension
commands in Zephyr v1.14 LTS as long as that is still alive. This
change is targeted towards Zephyr 2.0 users.)

This requires a bump in the shippable container and allows us to
simplify the west_commands test procedure.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

carlescufi commented on doc/guides/west/release-notes.rst:
```diff
@@ -1,6 +1,91 @@
+
+- No separate bootstrapper
+
+  In west v0.5.x, the program was split into two components, a bootstrapper and
```
Maybe you could link to the 1.14 doc here, where all the concepts described in this paragraph are described

mbolivar replied:
Thanks for bringing up the v1.14 documentation. I'll do that here, and send a separate PR to the v1.14 branch with the following changes as well:

- install west==0.5.8, making note that later versions may work, but it's not guaranteed
- `west init --mr v1.14.1` (since we've already cut v1.14.0, I guess that PR will go out with the next point release)

carlescufi replied:
Thanks, I've already reviewed the separate PR

carlescufi commented on doc/guides/west/release-notes.rst:
```diff
@@ -1,6 +1,91 @@
+  via PyPI and installed using ``pip3``.
+
+  This is similar to how Google's Repo tool works, but was determined to be a
+  mistake for west.
```
"mistake" is a strong word. It served its purpose (quick iteration), so I would rephrase this as "west is now stable enough to be entirely distributed through PyPI"

carlescufi commented on doc/guides/west/release-notes.rst:
```diff
@@ -1,6 +1,91 @@
+  not just extension commands.
+- The ``selfupdate`` command still exists for backwards compatibility, but
+  now simply exits after printing an error message.
+- Manifest syntax changes
```
Maybe mention that the changes are backwards compatible

mbolivar replied:
The syntax is compatible, but the semantics are not (project names must now be unique, which was previously not required), so I don't want to mislead people and would rather not write that here.

carlescufi commented on doc/guides/west/repo-tool.rst:
```diff
@@ -152,62 +148,12 @@ The following three source code topologies supported by west:
-<west-apis-west>` are also currently provided to extension commands by this
-repository. So that west can update itself, the built-in ``west update`` and
-``west selfupdate`` commands fetch and update the :file:`.west/west` repository.
+This distribution also includes a launcher binary, also named ``west``. When
```
This is actually not correct on Linux, where the "binary" is actually a Python script

mbolivar replied:
I will say "launcher executable"

carlescufi reviewed (changes requested):
Nice job! Just a few minor nitpicks on the documentation side of things

mbolivar commented on doc/CMakeLists.txt:
```diff
@@ -15,6 +15,11 @@ endif()
 file(TO_CMAKE_PATH "$ENV{ZEPHYR_BASE}" ZEPHYR_BASE)
 
 message(STATUS "Zephyr base: ${ZEPHYR_BASE}")
+if(DEFINED WEST)
```
We have it already:

```
-- Found west: /home/mbolivar/.virtualenvs/west-0.6/bin/west (found suitable version "0.6.0", minimum required is "0.5.6")
```

mbolivar commented:
@carlescufi thanks for review; comments addressed

dbkinder commented on doc/guides/west/build-flash-debug.rst:
```diff
@@ -463,13 +463,22 @@ determined by the imported subclasses of ``ZephyrBinaryRunner``.
+Hacking
+*******
+
+This section contains documentation the ``runners.core`` module used by the
```
```suggestion
This section documents the ``runners.core`` module used by the
```

dbkinder commented on doc/guides/west/release-notes.rst:
```diff
@@ -0,0 +1,116 @@
+  a per-installation clone. See `Multiple Repository Management in the v1.14
+  documentation`_ for more details.
+
+  This is similar to how Google's Repo tool works, and let west iterate quickly
```
```suggestion
  This is similar to how Google's Repo tool works, and lets west iterate quickly
```

dbkinder commented on doc/guides/west/repo-tool.rst:
```diff
@@ -152,62 +148,12 @@ The following three source code topologies supported by west:
-``west selfupdate`` commands fetch and update the :file:`.west/west` repository.
+This distribution also includes a launcher executable, also named ``west``. When
+west is installed, the launcher is placed by :file:`pip3` somewhere in the
+user's ``PATH``. This the command-line entry point.
```
```suggestion
user's ``PATH``. This is the command-line entry point.
```

dbkinder reviewed (changes requested):
Looks pretty good, thanks!  A few typos found...

mbolivar commented:
@dbkinder I've applied all your suggestions as-is. Thanks!

mbolivar commented:
Retesting with new CI container that has v0.6.0rc1 installed.

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -36,7 +36,7 @@ we also have :ref:`documentation for doing without it <no-west>`.)
 Bootstrap west
```
Rename this from Bootstrap to Install

mbolivar replied:
done

mbolivar-nordic committed:
scripts: un-break test_nrfjprog.py

The pytest.raises context manager is now returning an ExceptionInfo
whose str() doesn't contain the str() of the underlying exception
object. Take str(e.value) directly to make sure we're looking at the
exception string.

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>

mbolivar commented:
Rebased and fixed run_ci.sh now that west can no longer list itself.

vanwinkeljan commented on doc/guides/west/release-notes.rst:
```diff
@@ -0,0 +1,116 @@
+           - name: example-project-name
+             url: https://github.com/example/example-project
+
+    Project elements with ``url`` attributes set in this way may not also have
```
Would it be an option to add an attribute `remote-name` that can be combined with the attribute `remote`, or the default remote, to form the final URL? The current proposal forces to use the attribute `url` in case two or more projects share the same repo.

As an example for PR #17629 the current west 0.6 implementation would result in:

```
    - name: lvgl_v5
      revision: pull/1/head
      path: modules/lib/gui/lvgl_v5
      url: https://github.com/zephyrproject-rtos/lvgl
    - name: lvgl_v6
      revision: pull/2/head
      path: modules/lib/gui/lvgl_v6
       url: https://github.com/zephyrproject-rtos/lvgl
```
If the `remote-name` attribute would be added this could become:
```
    - name: lvgl_v5
      revision: pull/1/head
      path: modules/lib/gui/lvgl_v5
      remote-name: lvgl
    - name: lvgl_v6
      revision: pull/2/head
      path: modules/lib/gui/lvgl_v6
      remote-name: lvgl
```

mbolivar replied:
> If the remote-name attribute would be added this could become:

How? Sorry, I don't understand the proposal. The `remote` attribute is already the name of the remote, so I'm not sure what `remote-name` means.

```
    - name: lvgl_v5
      revision: pull/1/head
      path: modules/lib/gui/lvgl_v5
      remote-name: lvgl
```

I'm not sure what this is doing; can you please explain? Do you perhaps mean `repo-name` instead of `remote-name`?

> The current proposal forces to use the attribute url in case two or more projects share the same repo.

Right, but what's the issue with that?

mbolivar replied:
Also, would you mind taking this to the west issue tracker? This PR is about documenting the release.

vanwinkeljan replied:
> I'm not sure what this is doing; can you please explain? Do you perhaps mean `repo-name` instead of `remote-name`?

Yes, maybe that is a better naming for it, so `remote` + `repo-name` would form the URL.

> Right, but what's the issue with that?

Redundancy, is this not the reason why the remote concept was introduced?
In case more and more projects are introduced to support different API versions all of them need a hard coded `url`.

> Also, would you mind taking this to the west issue tracker? 

Ofcourse

mbolivar commented:
Kicking CI; there was a timeout (the test passes locally for me with west 0.6.0)

mbolivar commented:
Updated docs for 0.6.0rc3 with some additional changes based on feedback from @vanwinkeljan (for which, thanks!).

Once https://github.com/zephyrproject-rtos/ci-dockerfiles/pull/46 is merged and the container is tagged to 0.8rc2, we can run the documentation through CI one last time here, then cut the west release etc.

mbolivar commented:
Kicking CI again. @dbkinder could you please take another look at the docs? This is ready to merge from the code perspective.

dbkinder reviewed (approved):
docs LGTM, thanks!

mbolivar commented:
@nashif @MaureenHelm @galak are you good with this too?

galak commented on .shippable.yml:
```diff
@@ -21,7 +21,7 @@ build:
     pre_ci_boot:
         image_name: zephyrprojectrtos/ci
-        image_tag: v0.7
+        image_tag: v0.8.0-rc1
```
Made a v0.8.0 release with west 0.6.0

mbolivar replied:
I've updated this to use the v0.8.0 tag now that I see this is up on docker hub. Ready for this to merge once CI passes?

galak reviewed (commented):
Looks good to me, suggest to wait til 0.8.0 dockerimage gets published.

carlescufi commented:
@galak how far are we to be able to deploy 0.8.0?

carlescufi commented:
Tested with Docker Image 0.8.0
